src/emqx_client.erl

```diff
@@ -899,6 +987,85 @@ code_change(_Vsn, State, Data, _Extra) ->
 %% Internal functions
 %%------------------------------------------------------------------------------
 
+publish_process(?QOS_1, Packet = ?PUBLISH_PACKET(?QOS_1, PacketId), State = #state{auto_ack = AutoAck}) ->
+    _ = deliver(packet_to_msg(Packet), State),
+    case AutoAck of
+        true  -> send_puback(?PUBACK_PACKET(PacketId), State);
+        false -> {keep_state, State}
+    end;
+publish_process(?QOS_2, Packet = ?PUBLISH_PACKET(?QOS_2, PacketId),
+    State = #state{awaiting_rel = AwaitingRel}) ->
+    case send_puback(?PUBREC_PACKET(PacketId), State) of
+        {keep_state, NewState} ->
+            AwaitingRel1 = maps:put(PacketId, Packet, AwaitingRel),
+            {keep_state, NewState#state{awaiting_rel = AwaitingRel1}};
+        Stop -> Stop
+    end.
+
+response_process(ResponseInfo, _Payload) when is_list(ResponseInfo) ->
+    response_process(iolist_to_binary(ResponseInfo), _Payload);
+response_process(ResponseInfo, _Payload) when is_binary(ResponseInfo) ->
+    ResponseInfo;
+response_process(ResponseFun, Payload) when is_function(ResponseFun) ->
+    ResponseFun(Payload).
+
+shared_topic(Shared, Topic) ->
+    case Shared of
+        true ->
+            <<"$shared/", Topic/binary>>;
```

Comment:
a valid shared topic is `$shared/groupname/topicname`, this implementation will take the first portion in `Topic` as group name. And if `Topic` has only one portion, then `$shared/onlye-one-portiion` is invalid.